url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/generator/NetCoreGenerator.java

```diff
@@ -0,0 +1,1162 @@
+            idx += Link.LINK_SIZE;
+        }
+        //
+        // dont count the first weight, which is constant 1.
+        //
+        this.structure.weightsnum = (woff - 1);
+        //
+        // create reverted links. the weight indices are
+        // now in linksrev too. the reverted are sorted in
+        // source major order.
+        //
+        int[] linksrev = links.clone();
+        Link.sortSrcMaj(linksrev, Link.ORDER_ASC);
+        //
+        // swap src and dst in linksrev.
+        //
+        idx = 0;
+        for (int i = 0; i < linksnum; i++) {
+            final int src  = linksrev[idx + Link.IDX_SRC];
+            final int dst  = linksrev[idx + Link.IDX_DST];
+            //
+            linksrev[idx + Link.IDX_DST] = src;
+            linksrev[idx + Link.IDX_SRC] = dst;
+            //
+            idx += Link.LINK_SIZE;
+        }
+        this.structure.linksrev = linksrev;
+    }
+    
+    /**
+     * Configure cell arrays. Map predecessor and successor on each cell array.
+     */
+    private void setupArrays() {
+        //
+        // prepare arrays and sum indeg and outdeg.
+        //
+        CellArray[] arrays = new CellArray[this.arrays.size()];
+        for (int i = 0; i < arrays.length; i++) {
+            final CellArray a = this.arrays.get(i);
+            //
+            for (int j = a.cellslbd; j <= a.cellsubd; j++) {
+                a.indeg  += this.indeg[j];
+                a.outdeg += this.outdeg[j];
+            }
+            //
+            arrays[i] = a;
+        }
+        //
+        // forward:
+        // map arrays (pred) to links and count weights
+        //
+        int link = 0;
+        //
+        for (int i = 0; i < arrays.length; i++) {
+            //
+            final CellArray a = arrays[i];
+            if ((a.cellsnum <= 0) || (a.indeg == 0)) continue;
+            final int albd = a.cellslbd;
+            final int aubd = a.cellsubd;
+            //
+            int lbd  = -1;
+            int ctr  = 0;
+            //
+            // find links range begin by dst idx.
+            //
+            while (
+                    ((link + Link.IDX_DST) < this.structure.links.length) &&
+                    (this.structure.links[link + Link.IDX_DST] < albd)
+            ) {
+                link += Link.LINK_SIZE;
+            }
+            lbd = link;
+            //
+            // find links range end by dst idx.
+            //
+            while (
+                    ((link + Link.IDX_DST) < this.structure.links.length) &&
+                    (this.structure.links[link + Link.IDX_DST] <= aubd)
+            ) {
+                link += Link.LINK_SIZE;
+                ctr++;
+            }
+            //
+            a.predslbd = lbd;
+            a.predsnum = ctr;
+            //
+            // changed: 01.02.2012, Sebastian Otte
+            // a.predsubd = (a.predslbd + a.predsnum) - 1;
+            //
+            a.predsubd = link - Link.LINK_SIZE;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 321:358, 365:406.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.